### PR TITLE
Multi-gpu support

### PR DESCRIPTION
**Pull request by vbvg2008**
now FE can run on multi-gpu without code change

### CODE REVIEW

File: fastestimator/estimator/estimator.py [hunk 1]
```diff
@@ -140,23 +140,41 @@ def _warmup(self):
             for epoch in signature_epochs:
                 ds_iter = self.pipeline.dataset_schedule[mode].get_current_value(epoch)
                 batch = next(ds_iter)
-                global_batch_size = self.pipeline.get_global_batch_size(epoch)
+                if self.distribute_strategy and mode == "train":
```
Comment (TortoiseHam) [inline fastestimator/estimator/estimator.py]: do we still need to limit to train mode since we're doing the prediction concatenation now?

File: fastestimator/network/network.py [hunk 2]
```diff
@@ -85,27 +87,29 @@ def _get_signature_epoch(self, mode):
...
         ops = self.op_schedule[mode].get_current_value(epoch)
         model_list = self.model_schedule[mode].get_current_value(epoch)
-        loss_list = []
+        losses_epoch = []
         for model in model_list:
-            if model.loss_name not in loss_list:
-                loss_list.append(model.loss_name)
-        self.loss_list = loss_list
-        return ops, model_list, loss_list
+            if model.loss_name not in losses_epoch:
```
Comment (TortoiseHam) [inline fastestimator/network/network.py]: we might want to make losses_epoch a set rather than list to speed up the 'not in' check

File: fastestimator/network/network.py [hunk 3]
```diff
@@ -32,7 +34,7 @@ def __init__(self, ops):
         self.current_epoch_model = {}
         self.model = {}
         self.all_losses = []
-        self.loss_list = []
+        self.losses_epoch = []
```
Comment (TortoiseHam) [inline fastestimator/network/network.py]: this name might make more sense as epoch_losses to be more consistent with 'all_losses' from above

File: fastestimator/network/network.py [hunk 4]
```diff
@@ -128,9 +133,11 @@ def _forward(batch, state, ops):
...
-        for loss_key in loss_list:
-            loss = batch[loss_key]
+    def _reduce_loss(batch, global_batch_size, losses_epoch, warm_up):
+        reduced_loss = {}
+        for loss_name in losses_epoch:
+            element_wise_loss = batch[loss_name]
             if warm_up:
-                assert loss.shape.num_elements() > 1, "please make sure loss is element-wise loss"
-            batch[loss_key] = tf.reduce_sum(loss) / global_batch_size
+                assert element_wise_loss.shape.num_elements() > 1, "please make sure loss is element-wise loss"
```
Comment (TortoiseHam) [inline fastestimator/network/network.py]: what if batch size == 1?

File: fastestimator/pipeline/pipeline.py [hunk 5]
```diff
@@ -99,6 +100,8 @@ def prepare(self, distribute_strategy=None):
         for mode in self.mode_list:
             self._get_feature_name(mode)
             self._extract_dataset(mode)
+            if mode != "train":
+                distribute_strategy = None
```
Comment (TortoiseHam) [inline fastestimator/pipeline/pipeline.py]: as in earlier comment, can we not distribute the eval mode?

File: showcase/image_generation/dcgan_mnist.py [hunk 6]
```diff
@@ -92,10 +92,10 @@ def forward(self, data, state):
         return data
 
 
-def get_estimator(batch_size=256, epochs=10):
+def get_estimator(batch_size=256, epochs=50):
     # prepare data
-    (x_train, _), (x_eval, _) = tf.keras.datasets.mnist.load_data()
-    data = {"train": {"x": np.expand_dims(x_train, -1)}, "eval": {"x": np.expand_dims(x_eval, -1)}}
+    (x_train, _), (_, _) = tf.keras.datasets.mnist.load_data()
```
Comment (TortoiseHam) [inline showcase/image_generation/dcgan_mnist.py]: if we're not doing eval do we want to add that data into the train set?

File: fastestimator/network/network.py (hunk 3 above)
Comment (vbvg2008) [reply fastestimator/network/network.py]: fixed

File: fastestimator/network/network.py (hunk 2 above)
Comment (vbvg2008) [reply fastestimator/network/network.py]: I think the computational complexity for member check is the same.   Also, set is not iterable, which will create problems when we try to iterate through them.

File: showcase/image_generation/dcgan_mnist.py (hunk 6 above)
Comment (vbvg2008) [reply showcase/image_generation/dcgan_mnist.py]: I think we will do the same as tensorflow tutorial.   no need to add that.

File: fastestimator/network/network.py (hunk 2 above)
Comment (TortoiseHam) [reply fastestimator/network/network.py]: sets are iterable. also, unless python did sets totally wrong, membership check in set is O(1) vs in list O(n)

File: fastestimator/pipeline/pipeline.py (hunk 5 above)
Comment (vbvg2008) [reply fastestimator/pipeline/pipeline.py]: fixed

File: fastestimator/network/network.py (hunk 4 above)
Comment (vbvg2008) [reply fastestimator/network/network.py]: fixed
